### PR TITLE
Made many s3 client options configurable via spring values #1871

### LINKED ISSUE #1871: Provide possiblity for administrators to customize s3 client time out handling at deployment time
### Situation

Currently we use the AWS S3 client with the default setup and there is no way to define 

#### Defaults currently:

The class `com.amazonaws.ClientConfiguration` has currently following relevant default entries:
```java
    /** The default timeout for creating new connections. */
    public static final int DEFAULT_CONNECTION_TIMEOUT = 10 * 1000;

    /** The default timeout for reading from a connected socket. */
    public static final int DEFAULT_SOCKET_TIMEOUT = 50 * 1000;

    /**
     * The default timeout for a request. This is disabled by default.
     */
    public static final int DEFAULT_REQUEST_TIMEOUT = 0;

    /**
     * The default timeout for a request. This is disabled by default.
     */
    public static final int DEFAULT_CLIENT_EXECUTION_TIMEOUT = 0;

    /** The default max connection pool size. */
    public static final int DEFAULT_MAX_CONNECTIONS = 50;

    /**
     * The default expiration time (in milliseconds) for a connection in the connection pool.
     */
    public static final long DEFAULT_CONNECTION_TTL = -1;

    /**
     * The default maximum idle time (in milliseconds) for a connection in the connection pool.
     */
    public static final long DEFAULT_CONNECTION_MAX_IDLE_MILLIS = 60 * 1000;

    /**
     * The default time a connection can be idle in the connection pool before it must be validated that it's still open.
     */
    public static final int DEFAULT_VALIDATE_AFTER_INACTIVITY_MILLIS = 5 * 1000;


```
There are still additional setup parameters (retry etc.) but we do not handle those inside this issue.

### Wanted
The admins shall be able to configure via spring values (e.g. ENV variables) the s3 client setup.
This shall work with PDS and also with SecHub server applications.

## Solution
Enhance `S3Setup` with additional (optional) data.
Change `AwsS3JobStorageFactory` and use s3 setup.
We make also the signer override configurable (currently done by a static value)

### PR DESCRIPTION
**Pull request by de-jcup**
This PR
- closes #1871

### CODE REVIEW

File: sechub-pds/src/main/java/com/mercedesbenz/sechub/pds/storage/PDSS3PropertiesSetup.java [hunk 1]
```diff
@@ -27,6 +27,48 @@ public class PDSS3PropertiesSetup implements S3Setup {
     @Value("${sechub.pds.storage.s3.endpoint:" + UNDEFINED + "}") // we use undefined here. Will be used in isValid
     private String endpoint;
 
+    /* timeout */
+
+    @PDSMustBeDocumented(value = "S3 client timeout for creating new connections.", scope = "storage")
```
Comment (sven-dmlr) [inline sechub-pds/src/main/java/com/mercedesbenz/sechub/pds/storage/PDSS3PropertiesSetup.java:32]:
```suggestion
    @PDSMustBeDocumented(value = "S3 client timeout (in milliseconds) for creating new connections.", scope = "storage")
```

File: sechub-pds/src/main/java/com/mercedesbenz/sechub/pds/storage/PDSS3PropertiesSetup.java [hunk 2]
```diff
@@ -27,6 +27,48 @@ public class PDSS3PropertiesSetup implements S3Setup {
...
+    @Value("${sechub.pds.storage.s3.timeout.connection.milliseconds:" + S3Setup.DEFAULT_CONNECTION_TIMEOUT + "}")
+    private int connectionTimeoutInMilliseconds;
+
+    @PDSMustBeDocumented(value = "S3 client timeout in milliseconds for reading from a connected socket.", scope = "storage")
+    @Value("${sechub.pds.storage.s3.timeout.socket.milliseconds:" + S3Setup.DEFAULT_SOCKET_TIMEOUT + "}")
+    private int socketTimeoutInMilliseconds;
+
+    @PDSMustBeDocumented(value = "S3 client timeout for a request. 0 means it is disabled.", scope = "storage")
```
Comment (sven-dmlr) [inline sechub-pds/src/main/java/com/mercedesbenz/sechub/pds/storage/PDSS3PropertiesSetup.java:40]:
```suggestion
    @PDSMustBeDocumented(value = "S3 client timeout (in milliseconds) for a request. 0 means it is disabled.", scope = "storage")
```

File: sechub-pds/src/main/java/com/mercedesbenz/sechub/pds/storage/PDSS3PropertiesSetup.java [hunk 3]
```diff
@@ -27,6 +27,48 @@ public class PDSS3PropertiesSetup implements S3Setup {
...
+    @Value("${sechub.pds.storage.s3.timeout.request.milliseconds:" + S3Setup.DEFAULT_REQUEST_TIMEOUT + "}")
+    private int requestTimeoutInMilliseconds;
+
+    @PDSMustBeDocumented(value = "S3 client timeout for execution. 0 means it is disabled.", scope = "storage")
```
Comment (sven-dmlr) [inline sechub-pds/src/main/java/com/mercedesbenz/sechub/pds/storage/PDSS3PropertiesSetup.java:44]:
```suggestion
    @PDSMustBeDocumented(value = "S3 client timeout (in milliseconds) for execution. 0 means it is disabled.", scope = "storage")
```

File: sechub-pds/src/main/java/com/mercedesbenz/sechub/pds/storage/PDSS3PropertiesSetup.java [hunk 4]
```diff
@@ -27,6 +27,48 @@ public class PDSS3PropertiesSetup implements S3Setup {
...
+
+    @PDSMustBeDocumented(value = "S3 client expiration time (in milliseconds) for a connection in the connection pool. -1 means deactivated", scope = "storage")
+    @Value("${sechub.pds.storage.s3.connection.ttl.milliseconds:" + S3Setup.DEFAULT_CONNECTION_TTL + "}")
+    private long connectionTTLInMilliseconds;
+
+    @PDSMustBeDocumented(value = "S3 client maximum idle time (in milliseconds) for a connection in the connection pool.", scope = "storage")
+    @Value("${sechub.pds.storage.s3.connection.idle.max.milliseconds:" + S3Setup.DEFAULT_CONNECTION_MAX_IDLE_MILLIS + "}")
+    private long connectionMaxIdleInMilliseconds;
+
+    @PDSMustBeDocumented(value = "S3 client time a connection can be idle in the connection pool before it must be validated that it's still open.", scope = "storage")
```
Comment (sven-dmlr) [inline sechub-pds/src/main/java/com/mercedesbenz/sechub/pds/storage/PDSS3PropertiesSetup.java:62]:
```suggestion
    @PDSMustBeDocumented(value = "S3 client time (in milliseconds) a connection can be idle in the connection pool before it must be validated that it's still open.", scope = "storage")
```

File: sechub-pds/src/main/java/com/mercedesbenz/sechub/pds/storage/PDSS3PropertiesSetup.java [hunk 5]
```diff
@@ -27,6 +27,48 @@ public class PDSS3PropertiesSetup implements S3Setup {
...
+    @Value("${sechub.pds.storage.s3.connection.idle.validate.milliseconds:" + S3Setup.DEFAULT_VALIDATE_AFTER_INACTIVITY_MILLIS + "}")
+    private int validateAfterInactivityInMilliseconds;
+
+    /* signer */
+
+    @PDSMustBeDocumented(value = "S3 client time a connection can be idle in the connection pool before it must be validated that it's still open.", scope = "storage")
```
Comment (sven-dmlr) [inline sechub-pds/src/main/java/com/mercedesbenz/sechub/pds/storage/PDSS3PropertiesSetup.java:68]: The text is copied from line 62. Please fix.

File: sechub-pds/src/main/java/com/mercedesbenz/sechub/pds/storage/PDSS3PropertiesSetup.java [hunk 6]
```diff
@@ -27,6 +27,48 @@ public class PDSS3PropertiesSetup implements S3Setup {
     @Value("${sechub.pds.storage.s3.endpoint:" + UNDEFINED + "}") // we use undefined here. Will be used in isValid
     private String endpoint;
 
+    /* timeout */
+
+    @PDSMustBeDocumented(value = "S3 client timeout for creating new connections.", scope = "storage")
+    @Value("${sechub.pds.storage.s3.timeout.connection.milliseconds:" + S3Setup.DEFAULT_CONNECTION_TIMEOUT + "}")
+    private int connectionTimeoutInMilliseconds;
+
+    @PDSMustBeDocumented(value = "S3 client timeout in milliseconds for reading from a connected socket.", scope = "storage")
```
Comment (sven-dmlr) [inline sechub-pds/src/main/java/com/mercedesbenz/sechub/pds/storage/PDSS3PropertiesSetup.java:36]:
```suggestion
    @PDSMustBeDocumented(value = "S3 client timeout (in milliseconds) for reading from a connected socket.", scope = "storage")
```

File: sechub-shared-kernel/src/main/java/com/mercedesbenz/sechub/sharedkernel/storage/S3PropertiesSetup.java [hunk 7]
```diff
@@ -11,22 +11,64 @@
...
     private String bucketName;
 
-    @MustBeDocumented(value = "Defines the s3 endpoint - e.g. https://play.min.io", scope = "storage")
+    @MustBeDocumented(value = "Defines the S3 endpoint - e.g. https://play.min.io", scope = "storage")
     @Value("${sechub.storage.s3.endpoint:" + UNDEFINED + "}") // we use undefined here. Will be used in isValid
     private String endpoint;
 
+    /* timeout */
+
+    @MustBeDocumented(value = "S3 client timeout for creating new connections.", scope = "storage")
```
Comment (sven-dmlr) [inline sechub-shared-kernel/src/main/java/com/mercedesbenz/sechub/sharedkernel/storage/S3PropertiesSetup.java:32]:
```suggestion
    @MustBeDocumented(value = "S3 client timeout (in milliseconds) for creating new connections.", scope = "storage")
```

File: sechub-shared-kernel/src/main/java/com/mercedesbenz/sechub/sharedkernel/storage/S3PropertiesSetup.java [hunk 8]
```diff
@@ -11,22 +11,64 @@
...
+    @Value("${sechub.storage.s3.timeout.connection.milliseconds:" + S3Setup.DEFAULT_CONNECTION_TIMEOUT + "}")
+    private int connectionTimeoutInMilliseconds;
+
+    @MustBeDocumented(value = "S3 client timeout in milliseconds for reading from a connected socket.", scope = "storage")
```
Comment (sven-dmlr) [inline sechub-shared-kernel/src/main/java/com/mercedesbenz/sechub/sharedkernel/storage/S3PropertiesSetup.java:36]:
```suggestion
    @MustBeDocumented(value = "S3 client timeout (in milliseconds) for reading from a connected socket.", scope = "storage")
```

File: sechub-shared-kernel/src/main/java/com/mercedesbenz/sechub/sharedkernel/storage/S3PropertiesSetup.java [hunk 9]
```diff
@@ -11,22 +11,64 @@
...
+    @Value("${sechub.storage.s3.timeout.socket.milliseconds:" + S3Setup.DEFAULT_SOCKET_TIMEOUT + "}")
+    private int socketTimeoutInMilliseconds;
+
+    @MustBeDocumented(value = "S3 client timeout for a request. 0 means it is disabled.", scope = "storage")
```
Comment (sven-dmlr) [inline sechub-shared-kernel/src/main/java/com/mercedesbenz/sechub/sharedkernel/storage/S3PropertiesSetup.java:40]:
```suggestion
    @MustBeDocumented(value = "S3 client timeout (in milliseconds) for a request. 0 means it is disabled.", scope = "storage")
```

File: sechub-shared-kernel/src/main/java/com/mercedesbenz/sechub/sharedkernel/storage/S3PropertiesSetup.java [hunk 10]
```diff
@@ -11,22 +11,64 @@
...
+    @Value("${sechub.storage.s3.timeout.request.milliseconds:" + S3Setup.DEFAULT_REQUEST_TIMEOUT + "}")
+    private int requestTimeoutInMilliseconds;
+
+    @MustBeDocumented(value = "S3 client timeout for execution. 0 means it is disabled.", scope = "storage")
```
Comment (sven-dmlr) [inline sechub-shared-kernel/src/main/java/com/mercedesbenz/sechub/sharedkernel/storage/S3PropertiesSetup.java:44]:
```suggestion
    @MustBeDocumented(value = "S3 client timeout (in milliseconds) for execution. 0 means it is disabled.", scope = "storage")
```

File: sechub-shared-kernel/src/main/java/com/mercedesbenz/sechub/sharedkernel/storage/S3PropertiesSetup.java [hunk 11]
```diff
@@ -11,22 +11,64 @@
...
+
+    @MustBeDocumented(value = "S3 client expiration time (in milliseconds) for a connection in the connection pool. -1 means deactivated", scope = "storage")
+    @Value("${sechub.storage.s3.connection.ttl.milliseconds:" + S3Setup.DEFAULT_CONNECTION_TTL + "}")
+    private long connectionTTLInMilliseconds;
+
+    @MustBeDocumented(value = "S3 client maximum idle time (in milliseconds) for a connection in the connection pool.", scope = "storage")
+    @Value("${sechub.storage.s3.connection.idle.max.milliseconds:" + S3Setup.DEFAULT_CONNECTION_MAX_IDLE_MILLIS + "}")
+    private long connectionMaxIdleInMilliseconds;
+
+    @MustBeDocumented(value = "S3 client time a connection can be idle in the connection pool before it must be validated that it's still open.", scope = "storage")
```
Comment (sven-dmlr) [inline sechub-shared-kernel/src/main/java/com/mercedesbenz/sechub/sharedkernel/storage/S3PropertiesSetup.java:62]:
```suggestion
    @MustBeDocumented(value = "S3 client time (in milliseconds) a connection can be idle in the connection pool before it must be validated that it's still open.", scope = "storage")
```

File: sechub-shared-kernel/src/main/java/com/mercedesbenz/sechub/sharedkernel/storage/S3PropertiesSetup.java [hunk 12]
```diff
@@ -11,22 +11,64 @@
...
+    @Value("${sechub.storage.s3.connection.idle.validate.milliseconds:" + S3Setup.DEFAULT_VALIDATE_AFTER_INACTIVITY_MILLIS + "}")
+    private int validateAfterInactivityInMilliseconds;
+
+    /* signer */
+
+    @MustBeDocumented(value = "S3 client time a connection can be idle in the connection pool before it must be validated that it's still open.", scope = "storage")
```
Comment (sven-dmlr) [inline sechub-shared-kernel/src/main/java/com/mercedesbenz/sechub/sharedkernel/storage/S3PropertiesSetup.java:68]: The text is copied from line 62. Please fix.

**Review by sven-dmlr — CHANGES_REQUESTED** [review]
Looks good :+1: 

* Please check my comments about the documentation texts.

* It seems that `sharedkernel/storage/S3PropertiesSetup.java` and `pds/storage/PDSS3PropertiesSetup.java` are redundant. Are there good reasons for this?

**Comment by de-jcup** [comment]
> Looks good +1
> 
>     * Please check my comments about the documentation texts.
> 
>     * It seems that `sharedkernel/storage/S3PropertiesSetup.java` and `pds/storage/PDSS3PropertiesSetup.java` are redundant. Are there good reasons for this?

I changed the documentation parts with last commit

About the redundancy ... yes it is very similar, but we need this 
- we have different keys for the values (e.g. "sechub.pds.storage... " versus "sechub.storage...."
- PDS and SecHub are different products which do not communicate directly or share a domain etc. The only shared 
  parts inside the "common-*" libraries which have no dependency to spring/spring boot but we need spring values to 
  inject, so it cannot be shared.

**Review by sven-dmlr — APPROVED** [review]

**Pull request merged**